test/distributed/_composable/test_fully_shard.py

```diff
@@ -444,5 +444,53 @@ def test_float16_on_one_submodule(self):
             self.assertEqual(forward_inputs[c2].dtype, torch.float16)
 
 
+class TestFSDPModelCheckpointing(FSDPTest):
+    """Tests composable FSDP model checkpointing."""
+
+    @property
+    def world_size(self) -> int:
+        return 2
+
+    @skip_if_lt_x_gpu(2)
+    def test_state_dict_root_fully_shard(self):
+        """
+        Tests that the full state dict saved from a module with ``fully_shard``
+        applied to the global root matches that of an equivalent local module.
+        """
+        local_model = CompositeParamModel(device=torch.device("cuda"))
+        composable_module = copy.deepcopy(local_model)
+        fully_shard(composable_module, policy=ModuleWrapPolicy({UnitModule}))
+        local_sd = local_model.state_dict()
+        composable_sd = composable_module.state_dict()
+        self._check_state_dict_parity(local_sd, composable_sd)
+
+    @skip_if_lt_x_gpu(2)
+    def test_state_dict_submodule_fully_shard(self):
+        """
+        Tests that the full state dict saved from a module with ``fully_shard``
+        applied on submodules matches that of an equivalent local module.
+        """
+        local_model = CompositeParamModel(device=torch.device("cuda"))
+        composable_module = copy.deepcopy(local_model)
+        fully_shard(composable_module.u1)
+        fully_shard(composable_module.u2)
+        local_sd = local_model.state_dict()
+        composable_sd = composable_module.state_dict()
+        self._check_state_dict_parity(local_sd, composable_sd)
+
+    def _check_state_dict_parity(self, local_sd: Dict, composable_sd: Dict):
+        """Checks that ``local_sd`` and ``composable_sd`` are the same."""
+        # Check that all keys match
+        for k1, k2 in zip(composable_sd.keys(), local_sd.keys()):
+            self.assertEqual(k1, k2)
+        # Check that all values match
+        for v1, v2 in zip(composable_sd.values(), local_sd.values()):
+            self.assertEqual(v1.shape, v2.shape)
+        # Check that all keys and values are aligned
+        for (k1, v1), (k2, v2) in zip(composable_sd.items(), local_sd.items()):
```

Comment:
doesn't this encapsulate the above 2 checks?

Reply:
Yes, but the error message will be better. I followed your unit test 😃